particle-iot/device-os
# Particle.connected() should return true only after handshake messages are acknowledged

### Problem

`Particle.connected()` returns `true` immediately after the session is resumed, even if the session data became invalid for some reason and the device will in fact never receive ACKs for its handshake messages until the next connection attempt.

### Solution

1. Wait for all system- and protocol-level handshake messages to be acknowledged before changing the cloud connection state.
2. Make `Particle.publish()` and other cloud functions in the Wiring API fail early if the device is not connected to the cloud.

### Steps to Test

1. Run the communication unit tests.
2. Flash the example app and verify via the logs that `Particle.connected()` changes its return value only after all expected ACKs are received from the server.

### Example App

```cpp
#include "application.h"

SYSTEM_MODE(MANUAL)

namespace {

const SerialLogHandler logHandler(LOG_LEVEL_INFO);

void cloudStatusChanged(system_event_t event, int param) {
    if (param == cloud_status_connected) {
        Log.info("cloud_status_connected");
    }
}

bool lastConnected = true;

} // namespace

void setup() {
    System.on(cloud_status, cloudStatusChanged);
    Particle.connect();
}

void loop() {
    const bool connected = Particle.connected();
    if (connected != lastConnected) {
        Log.info("Particle.connected(): %d", (int)connected);
        lastConnected = connected;
    }
    Particle.process();
}
```

### References

- [ch32340]
